jxl/src/entropy_coding/ans.rs

```diff
@@ -2,6 +2,410 @@
 //
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
+//
+// Originally written for jxl-oxide.
+
+use crate::bit_reader::BitReader;
+use crate::error::{Error, Result};
+
+#[derive(Debug)]
+struct AnsHistogram {
+    buckets: Vec<Bucket>,
+    log_bucket_size: usize,
+    bucket_mask: u32,
+    // For optimizing fast-lossless case.
+    #[allow(unused)]
+    single_symbol: Option<u32>,
+}
+
+// log_alphabet_size <= 8 and log_bucket_size <= 7, so u8 is sufficient for symbols and cutoffs.
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+struct Bucket {
+    alias_symbol: u8,
+    alias_cutoff: u8,
+    dist: u16,
+    alias_offset: u16,
+    alias_dist_xor: u16,
+}
+
+impl AnsHistogram {
+    fn decode_dist_binary(br: &mut BitReader, dist: &mut [u16]) -> Result<usize> {
+        let table_size = dist.len();
+
+        let v0 = Self::read_u8(br)? as usize;
+        let v1 = Self::read_u8(br)? as usize;
+        if v0 == v1 {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let alphabet_size = v0.max(v1) + 1;
+        if alphabet_size > table_size {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let prob = br.read(12)? as u16;
+        dist[v0] = prob;
+        dist[v1] = (1u16 << 12) - prob;
+
+        Ok(alphabet_size)
+    }
+
+    fn decode_dist_unary(br: &mut BitReader, dist: &mut [u16]) -> Result<usize> {
+        let table_size = dist.len();
+
+        let val = Self::read_u8(br)? as usize;
+        let alphabet_size = val + 1;
+        if alphabet_size > table_size {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        dist[val] = 1 << 12;
+
+        Ok(alphabet_size)
+    }
+
+    fn decode_dist_evenly_distributed(br: &mut BitReader, dist: &mut [u16]) -> Result<usize> {
+        let table_size = dist.len();
+
+        let alphabet_size = Self::read_u8(br)? as usize + 1;
+        if alphabet_size > table_size {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let base = (1usize << 12) / alphabet_size;
+        let remainder = (1usize << 12) % alphabet_size;
+        dist[0..remainder].fill(base as u16 + 1);
+        dist[remainder..alphabet_size].fill(base as u16);
+
+        Ok(alphabet_size)
+    }
+
+    fn decode_dist_complex(br: &mut BitReader, dist: &mut [u16]) -> Result<usize> {
+        let table_size = dist.len();
+
+        let mut len = 0usize;
+        while len < 3 {
+            if br.read(1)? != 0 {
+                len += 1;
+            } else {
+                break;
+            }
+        }
+
+        let shift = (br.read(len)? + (1 << len) - 1) as i16;
+        if shift > 13 {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let alphabet_size = Self::read_u8(br)? as usize + 3;
+        if alphabet_size > table_size {
+            return Err(Error::InvalidAnsHistogram);
+        }
+
+        let mut repeat_ranges = Vec::new();
```

Comment:
This could be a `[..; MAX_ALPHABET_SIZE]`, instead of a vector, right?
I suspect it wouldn't change much performance wise, but leaving a todo in case it does might be nice.